Add functionality to write turn lookups and read turn penalty files

lbud opened the pull request:
Reopened from https://github.com/Project-OSRM/osrm-backend/pull/2291.

lbud commented:
k @danpat @TheMarex @MoKob and/or @daniel-j-h 🌴 have at it

danpat commented on src/extractor/edge_based_graph_factory.cpp:
```diff
@@ -431,6 +431,33 @@ void EdgeBasedGraphFactory::GenerateEdgeExpandedEdges(
+                        isCompressed
+                            ? m_node_info_list[m_compressed_edge_container.GetLastEdgeSourceID(
+                                  edge_from_u)]
+                            : m_node_info_list[node_u];
```
The behaviour here looks right, but I think the wording could be changed.  I think if I were to come back to this in a few months, `isCompressed` and `ContainsCompressedGeometries` might confuse me, as technically, all edges are in the `CompressedGeometryContainer`.

How about `isTrivial` and `isTrivial()`, and the some comments in the `CompressdGeometryContainer` to explain what `Trivial` means in this context.

TheMarex commented on include/engine/routing_algorithms/routing_base.hpp:
```diff
@@ -301,8 +301,6 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
 
                 BOOST_ASSERT(weight_vector.size() == id_vector.size());
-                // ed.distance should be total_weight + penalties (turn, stop, etc)
-                BOOST_ASSERT(ed.distance >= total_weight);
```
Why was this assertion removed? I don't see why this should be wrong unless there are negative penalties?

danpat replied:
@TheMarex yes, that's exactly the case this was removed to support.

The situation where this might arise would be where segment speeds are being calculated using GPS probe data.  At an intersection where the majority of the traffic makes a turn, the average speed will be primarily that of turning traffic.  However, if the intersection also allows unimpeded straight-through movements, a negative "turn" value can be used to correct traversal time for traffic that proceeds straight and unimpeded.  This type of situation is reasonably common on the outskirts of urban areas across the US, example: http://www.openstreetmap.org/#map=18/38.79196/-121.35847

Really though, it would apply in any situation where an average measurement does not represent the fastest traffic through an intersection, and a "speedup" needs to be applied for some movements.

TheMarex commented on src/contractor/contractor.cpp:
```diff
@@ -40,6 +41,16 @@ template <> struct hash<std::pair<OSMNodeID, OSMNodeID>>
+    {
+        return (static_cast<uint64_t>(std::get<0>(k)) << 24) ^
+               (static_cast<uint64_t>(std::get<1>(k)) << 12) ^
+               (static_cast<uint64_t>(std::get<2>(k)));
```
Instead of an own hash functio you should consider using `boost:hash_combine`:

http://www.boost.org/doc/libs/1_55_0/doc/html/hash/combine.html

TheMarex commented on src/extractor/compressed_edge_container.cpp:
```diff
@@ -250,16 +250,25 @@ CompressedEdgeContainer::GetBucketReference(const EdgeID edge_id) const
+NodeID CompressedEdgeContainer::GetLastEdgeTargetID(const EdgeID edge_id) const
+{
+    const auto &bucket = GetBucketReference(edge_id);
+    return bucket[bucket.size() - 1].node_id;
```
Use `bucket.back().node_id` instread.

TheMarex commented on src/extractor/compressed_edge_container.cpp:
```diff
@@ -250,16 +250,25 @@ CompressedEdgeContainer::GetBucketReference(const EdgeID edge_id) const
 NodeID CompressedEdgeContainer::GetLastEdgeSourceID(const EdgeID edge_id) const
 {
     const auto &bucket = GetBucketReference(edge_id);
-    BOOST_ASSERT(bucket.size() >= 2);
```
Don't remove but correct assertion for uncompressed case: `BOOST_ASSERT(bucket.size() >= 1);`

lbud commented:
Added a few comments for clarity, replaced the removed assertions, renamed the new compressed geometry fns/vars to `isTrivial`/`IsTrivial()`, and rebased on master. As long as https://github.com/Project-OSRM/osrm-backend/pull/2306#discussion_r60853416 is okay I think this is ready — 🚦✅ and I'll squash and merge.

TheMarex commented:
@lbud tests on travis seem to be failing. Otherwise :+1: for merging if that is fixed.

lbud commented on include/engine/routing_algorithms/routing_base.hpp:
```diff
@@ -397,8 +396,17 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
+            // node to the first turn would be the same as from end to end of a segment,
+            // which is obviously incorrect and not ideal...
+            unpacked_path.front().duration_until_turn =
+                std::max(unpacked_path.front().duration_until_turn - source_weight, 1);
```
@TheMarex would love another set of 👀  on this — it feels like maybe sort of a hacky solution to potentially negative edge durations, but… ??
cc @danpat

TheMarex replied:
Actually, this explains why the test fails above. If the phantom node snapps to a segment end point the result of this used to be `0`. using `0` instead of `1` here should yield the expected behavior.

lbud commented on src/contractor/contractor.cpp:
```diff
@@ -494,7 +545,34 @@ std::size_t Contractor::LoadEdgeExpandedGraph(
+                        << " is too negative: clamping turn weight to " << original_node_count;
+                }
+
+                inbuffer.weight = std::max(new_turn_weight + new_weight, original_node_count);
```
@TheMarex @daniel-j-h @MoKob @danpat 

I wasn't sure whether to clamp this to `original_node_count` (same as `num_osm_nodes` before it gets modified), `original_node_count * 10`, or `1` (or something else)… 🍝 👾  help?

lbud commented:
Travis tests were failing because I was testing a debug assertion in cucumber, so all release builds weren't erroring where I expected they should — which exposed the issue that we really needed to catch any negative edge weight regardless of build type, hence https://github.com/Project-OSRM/osrm-backend/pull/2306/commits/663c28ee6b92b90f23eec5b6f0312aedd2c513ff. This raised some additional questions which I've added in comments here.

TheMarex commented:
@lbud can you rebase again on `master`? Seems like the travis breakages are mostly due to me accidentally breaking master.

lbud commented:
👌 @TheMarex just rebased after renaming `original_node_count` to `compressed_edge_nodes`

lbud commented:
Just forced pushed after squash/rebase and then I'll merge.

MoKob commented on src/extractor/compressed_edge_container.cpp:
```diff
@@ -250,16 +250,32 @@ CompressedEdgeContainer::GetBucketReference(const EdgeID edge_id) const
-    BOOST_ASSERT(bucket.size() >= 2);
+    BOOST_ASSERT(bucket.size() >= 1);
     return bucket[bucket.size() - 2].node_id;
 }
```
`bucket.size() >=1` and access to `bucket.size()-2` seems like a bad idea?

TheMarex replied:
Derp. That might have been my ill-advised comment to consider that non-compressed geometries. Yes this should be `2`.
